### PR TITLE
Refresher_spec: Add deletions tests

### PR DESCRIPTION
**Pull request by zakiva**
Adding deletions tests to `refresher_spec.rb`, depends on: https://github.com/ManageIQ/manageiq/pull/15182

### CODE REVIEW

**Comment by zakiva** [comment]
@simon3z @moolitayer @cben Please review

**Comment by moolitayer** [comment]
@zakiva as discussed 
- please move inventory generation to use templates for creation & deletion
- add the templates to our source code and document the commands needed to recreate the inventory

**Comment by simon3z** [comment]
@zakiva @moolitayer I haven't reviewed but can you make sure that this is all relevant just for OpenShift? (If there's anything related to Kubernetes it belongs in the other repo). Thanks.

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb [hunk 1]
```diff
@@ -127,6 +127,228 @@ def normal_refresh
...
+    assert_specific_container_definitions_disconnected
+    assert_specific_containers_disconnected
+  end
+
+  def assert_specific_projects_disconnected
+    @project0.reload
+    expect(@project0).not_to be_nil
+    expect(@project0.ext_management_system).to be_nil
+    expect(@project0.deleted_on).not_to be_nil
+    expect(@project0.old_ems_id).to eq(@ems.id)
```
Comment (cben) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: Above 4 lines will change with https://github.com/ManageIQ/manageiq/pull/15251
They repeat for several objects here, so I'd extract them to an `assert_disconnected(@project0)` helper.

**Comment by cben** [comment]
I think you'll also need to change (or just remove?) these:
https://github.com/ManageIQ/manageiq/blob/e31348d96e912a13fed39a2d4762273437541457/app/models/manageiq/providers/container_manager.rb#L25-L28

File: spec/template.yml [hunk 2]
```diff
@@ -0,0 +1,183 @@
+apiVersion: v1
```
Comment (zakiva) [inline spec/template.yml]: @moolitayer any better place to locate this file?

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb [hunk 3]
```diff
@@ -127,6 +126,164 @@ def normal_refresh
...
+    expect(ContainerTemplate.find_by(:name => "my-template", :container_project => @project2)
+             .container_template_parameters.count).to eq(1)
+  end
+
+  def assert_table_counts_after_deletions
+    expect(ContainerProject.count).to eq(8)
+    expect(ContainerGroup.count).to eq(10)
+    expect(Container.count).to eq(10)
+    expect(ContainerDefinition.count).to eq(10)
+    expect(ContainerImage.count).to eq(44) # BUG?
```
Comment (zakiva) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: I don't know what is the expected behavior here. 
We save "2 kinds" of images in the DB: 
1) Images that were imported into the OpenShfit cluster which we get as all other resources in the inventory. 
2) Images that were not imported to the cluster, but are in use by one or more containers. 
If an image of the first kind is deleted in OpenShift we disconnect it on our side (do not delete it). However, deletion of all containers that run an image of the second kind causes a deletion of that image from the DB (not disconnect). 
cc @simon3z @enoodle

File: spec/template.yml [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+          kind: "ImageStreamTag"
+          name: "ruby-20-centos7:latest"
+    output: 
+      to:
+        kind: "ImageStreamTag"
+        name: "origin-ruby-sample:latest"
+    postCommit: 
+        script: "bundle exec rake test"
+- apiVersion: v1
+  kind: "Build"
```
Comment (zakiva) [inline spec/template.yml]: @cben This should cover https://github.com/ManageIQ/manageiq-providers-openshift/pull/19/files#diff-c7cbabc2b2e9df5cbda6d2ecc5afecc2R292 let me know if you want to test specific attributes/relations here.

**Comment by zakiva** [comment]
> @zakiva @moolitayer I haven't reviewed but can you make sure that this is all relevant just for OpenShift? (If there's anything related to Kubernetes it belongs in the other repo). Thanks.

@simon3z Some of the objects we test here are not "just for OpenShit", and in theory may be moved to the k8s repo. However, I'm not sure what's the benefit of testing things also there. I think we should keep it here as I'm using a template to create the objects and oc commands to delete them (OpenShift specific). Also, I think it will be easier to maintain tests and cassettes only here instead of having different cassettes for k8s and openshift in the different repos.

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb (hunk 1 above)
Comment (zakiva) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: done

File: spec/template.yml (hunk 2 above)
Comment (moolitayer) [inline spec/template.yml]: ```spec/vcr_cassettes/manageiq/providers/openshift/container_manager/``` is a good start. Give it a descriptive name so we know this template is used to create templates.

**Comment by moolitayer** [comment]
> @simon3z Some of the objects we test here are not "just for OpenShit", and in theory may be moved to the k8s repo. However, I'm not sure what's the benefit of testing things also there. I think we should keep it here as I'm using a template to create the objects and oc commands to delete them (OpenShift specific). Also, I think it will be easier to maintain tests and cassettes only here instead of having different cassettes for k8s and openshift in the different repos.
> BTW, we already have here (refresher_spec.rb) some tests for k8s entities.

We definitely need k8s tests in k8s repo and OSH tests in OSH repo, other wise our continues delivery is brittle. It is less important to me where the template lives - can be here. Please make sure that test code is split between the repositories - only that way we can detect breaking changes.

**Comment by zakiva** [comment]
> We definitely need k8s tests in k8s repo and OSH tests in OSH repo, other wise our continues delivery is brittle. It is less important to me where the template lives - can be here. Please make sure that test code is split between the repositories - only that way we can detect breaking changes.

I agree with that, but it may raise some dependency issues. If we want to keep the approach of creating the objects out of a template we'll have to use openshift provider and then save the cassette in this repo, then the tests that are k8s-related will be in the other repo and will use this cassette? 
Also, the initial purpose of these tests was to ensure that we delete/archive entities correctly, but the code that does that is in `save_inventory_container.rb` in the main repo, thus changes in that code in the future will not break any tests in that repo. So, maybe the solution will be to move the tests there? (same goes for existing tests in the `refresher_specs.rb` files in both repos).

**Comment by moolitayer** [comment]
> I agree with that, but it may raise some dependency issues. If we want to keep the approach of creating the objects out of a template we'll have to use openshift provider and then save the cassette in this repo, then the tests that are k8s-related will be in the other repo and will use this cassette?

The purpose of the template is to ease our life and unfortunately we have no kubernetes template. As I've said i'm fine with having one template here OR two templates in both repos. Do you have another a different idea?


> Also, the initial purpose of these tests was to ensure that we delete/archive entities correctly, but the code that does that is in save_inventory_container.rb in the main repo, thus changes in that code in the future will not break any tests in that repo. So, maybe the solution will be to move the tests there? (same goes for existing tests in the refresher_specs.rb files in both repos).

The fact that our inventory refresh implementation is split between core/plugin repos is not ideal.
This is the point to bring @cben into the discussion and make sure we are all aligned.

**Comment by moolitayer** [comment]
Come to think of it, as it is designed now save_inventory_container is testing the general container entities and should be in the main repo. After discussing this with @zakiva we decided
- Have these tests here as integration tests testing both the parsing code in this repo and the core infra in save_inventory_container
- Look into adding unit tests in manageiq/manageiq for save_inventory_container but only after we understand from @cben or @Ladas  what will be the design post graph refresh

**Comment by cben** [comment]
New graph refresh code will live in k8s, only legacy save_inventory_container
remains in core.
This will reduce CI issues.
(There is of course graph infrastracture in core, but there is nothing
container-specific there.)

I don't think we'll add any *unit* tests for save_inventory_container; if
anything is not covered we should add integration (refresher) tests to
cover both legacy & graph refresh.


2017-06-13 13:29 GMT+03:00 Mooli Tayer <notifications@github.com>:

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb [hunk 5]
```diff
@@ -127,6 +127,115 @@ def normal_refresh
...
+                         :match_requests_on => [:path,]) do # , :record => :new_episodes) do
+          EmsRefresh.refresh(@ems)
+        end
+      end
+
+      it "archives objects" do
+        expect(ContainerProject.count).to eq(8)
+        expect(ContainerProject.where(:deleted_on => nil).count).to eq(7)
+        expect(ContainerImage.count).to eq(43) # should be 44
+        expect(ContainerImage.where(:deleted_on => nil).count).to eq(43) # should be 44
```
Comment (cben) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: interesting.  do you know why?
consider extracting these to a `pending` test (actually asserting 44).

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb [hunk 6]
```diff
@@ -127,6 +127,115 @@ def normal_refresh
...
+      end
+
+      it "removes the deleted objects from the DB" do
+        expect(ContainerRoute.count).to eq(4)
+        expect(ContainerTemplate.count).to eq(28)
+        expect(ContainerReplicator.count).to eq(8)
+        expect(ContainerBuild.count).to eq(1)
+        expect(ContainerBuildPod.count).to eq(1)
+        expect(CustomAttribute.count).to eq(523)
```
Comment (cben) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: Travis fails saying it's actually 524.
It's not really possible to review this number :-)
Perhaps we can add assertion before deleting that the objects you're gonna delete had D attributes, and now the number is 532 - D - 1 (one for the label of the non-destroyed route).  Not blocker, we'll keep improving these with time.

Comment (zakiva) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: Travis will pass once https://github.com/ManageIQ/manageiq/pull/15182 will be merged

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb (hunk 5 above)
Comment (zakiva) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: yeah I mentioned it in a comment before but it is hidden writing it again: 

I don't know what is the expected behavior here.
We save "2 kinds" of images in the DB:

Images that were imported into the OpenShfit cluster which we get as all other resources in the inventory.
Images that were not imported to the cluster, but are in use by one or more containers.
If an image of the first kind is deleted in OpenShift we disconnect it on our side (do not delete it). However, deletion of all containers that run an image of the second kind causes a deletion of that image from the DB (not disconnect).
cc @simon3z @enoodle

**Comment by zakiva** [comment]
> Very nice 
> 
> Is anything blocking this beside the kubernetes/openshift split?
> I kinda feel we shot ourselves in the foot with having 2 repos, and cutting this test will just be an exercise in suffering for the sake of suffering :-(
> 
> I vote let's please merge this for now.
> => Sometimes this will break openshift repo because of k8s repo commits. But that's true anyway.

@cben Yeah I discussed it with @moolitayer we agreed on having these tests as they are now plus adding the k8s entities tests in the other repo. I'm working on it right now.

**Comment by simon3z** [comment]
@moolitayer this PR still shows that you requested changes. Review and feel free to merge if it's OK with you.

**Comment by cben** [comment]
@moolitayer if you still have rspec style comments, can we do them in later PR?

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb (hunk 3 above)
Comment (enoodle) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: It is a little late now, but @zeari Do you know if the this the expected behavior?

**Pull request merged**